wenet/transformer/attention.py

```diff
@@ -32,17 +35,36 @@ class MultiHeadedAttention(nn.Module):
         dropout_rate (float): Dropout rate.
 
     """
-    def __init__(self, n_head: int, n_feat: int, dropout_rate: float):
+    def __init__(self, n_head: int, n_feat: int, dropout_rate: float,
+                 lora_list: List[str] = None, lora_rank: int = 8,
+                 lora_alpha: int = 8, lora_dropout: float = 0.0):
         """Construct an MultiHeadedAttention object."""
         super().__init__()
         assert n_feat % n_head == 0
         # We assume d_v always equals d_k
         self.d_k = n_feat // n_head
         self.h = n_head
-        self.linear_q = nn.Linear(n_feat, n_feat)
-        self.linear_k = nn.Linear(n_feat, n_feat)
-        self.linear_v = nn.Linear(n_feat, n_feat)
-        self.linear_out = nn.Linear(n_feat, n_feat)
+        if lora_list is not None:
+            self.linear_out = lora.Linear(n_feat, n_feat, r=lora_rank,
+                                          lora_alpha=lora_alpha,
+                                          lora_dropout=lora_dropout) \
+                              if "o" in lora_list else nn.Linear(n_feat, n_feat)
+
+            lora_qkv_dict = {"q": "q" in lora_list, "k": "k" in lora_list,
+                             "v": "v" in lora_list}
+
+            for key, value in lora_qkv_dict.items():
+                setattr(self, f"linear_{key}",
+                        lora.MergedLinear(n_feat, n_feat, r=lora_rank,
```

Comment:
I don't quite understand. Could you please help explain why QKV consists of three separate dense layers, yet requires three MergedLinear layers?

Reply:
The LoRA method will freeze the original parameters, and a parallel branch of q,k,v,linear_out will add to it, when inferring mode, the branch parameters will merge into the net: Wx=(W+BA)x

Reply:
What I mean is that here, three dense layers are separated. Why not directly use `lora.Linear` and instead use three `lora.MergedLinear`?

Reply:
yes, u r right, i will modify this to make it clean, thx